sys: random: fortuna csprng

basilfx opened the pull request:
This PR adds support for the [Fortuna Computationally Secure PRNG](https://www.schneier.com/academic/paperfiles/fortuna.pdf). I had been working on this because of [this application note](https://www.silabs.com/Support%20Documents/TechnicalDocs/AN0806.pdf). I have based this on [this code](https://github.com/blin00/os/blob/master/src/random.c), but fully refactored it for RIOT-OS, while comparing it to the textbook implementation.

This is not your daily PRNG, as it is memory hungry. The internal state requires N \* `sha256_context_t`, where N is typically >= 16, or >= 2.5KB RAM. I wasn't sure to PR this, but after the summit I thought: why not? After all, it fits perfectly in the discussion of #5659. 

It is a WIP, because the reference implementation needs AES-256 (#4798, #5429). Now it uses AES-128, which works fine. Furthermore, initialization of the PRNG requires a seed >> 32 bits, so I cheat a bit. #4294 would help.

I'm planning to do some offline testing, eventually.

BytesGalore commented on Makefile.dep:
```diff
@@ -564,6 +564,14 @@ ifneq (,$(filter random,$(USEMODULE)))
+        USEMODULE += hashes
+        USEMODULE += crypto
+        USEMODULE += xtimer
+        CFLAGS += -DCRYPTO_AES 
```
I wonder if it would be more intuitive to set the CFLAG in `sys/random/Makefile`,
since here we only handle dependencies.
Beside, please remove the extra whitespace behind the flag.

PeterKietzmann replied:
How about sys/Makefile.dep?

BytesGalore commented on sys/random/fortuna/fortuna.c:
```diff
@@ -0,0 +1,173 @@
+        size_t len = 0;
+
+        for (size_t i = 0; i < FORTUNA_POOLS; i++) {
+            if (state->reseeds | (1 << i)) {
```
isn't this always true?

basilfx replied:
Nope, `state->reseeds` increments over time: this check tries to balance the use of each pool IIRC.

BytesGalore commented on tests/prng_tools/main.c:
```diff
@@ -0,0 +1,112 @@
+    uint64_t total = 0;
+
+    for (int i = 0; i < bins; i++) {
+        total = total + values[i];
```
how about `+=`

BytesGalore commented on tests/prng_tools/main.c:
```diff
@@ -0,0 +1,112 @@
+        total = total + values[i];
+    }
+
+    /* if total is zero, it would yield a division by zero */
```
then I propose to `assert` it here

basilfx replied:
IMHO `assert` would be a program error, where this case is a likely scenario (for a bad RNG).

BytesGalore commented:
nice one, seems to distribute the bits quite equally :+1:

basilfx commented on Makefile.dep:
```diff
@@ -564,6 +564,14 @@ ifneq (,$(filter random,$(USEMODULE)))
+        USEMODULE += hashes
+        USEMODULE += crypto
+        USEMODULE += xtimer
+        CFLAGS += -DCRYPTO_AES 
```
I think the `CFLAGS` is ugly here, but it's the only way to tell the `crypto` module to include AES. It's WIP after all :-)

basilfx commented:
Fixed few comments + rebase. Still WIP (but not abandoned).

basilfx commented:
For the record, this 33c3 talk may be of interest for this PR: [Wheel of Fortune - Analyzing Embedded OS Random Number Generators](https://media.ccc.de/v/33c3-7949-wheel_of_fortune).

miri64 commented:
For now: needs rebase.

basilfx commented:
Rebased.

basilfx commented:
I have updated the PR and I have re-checked my implementation with the [textbook](https://www.schneier.com/academic/paperfiles/fortuna.pdf) and couldn't spot any mistakes.

I have exposed more configuration options as defines and set them to recommended defaults (which is not memory friendly, but can be overridden).

I have removed the WIP label since I think this implementation is ready. It still depend on AES-256, but it works fine with AES-128 at the expense of some security. I don't fully understand the mathematical details, but having a 256-bit key instead of 128-bit reduces the chance of selecting a permutation of input -> outputs where collisions occur (please correct me!). That said, if AES-256 gets contributed, then it is a matter of changing one define.

basilfx commented:
I decided to take out the testing tools and move it to #8182, so back to WIP and waiting for pr.

basilfx commented:
Still working on this one. I've removed `rng_tools`, which was introduced using #8182. I still have to fix one remaining issue to seed the generator on start.

PeterKietzmann commented:
@basilfx thanks for the heads up. What exactly is the issue with seeding? Seeding this kind of generators is still a somehow open issue that also applies to other RNGs. Can we address it in a separate issue/PR?

basilfx commented:
I'm currently stuck on [this](https://github.com/basilfx/RIOT/blob/d80210987953a06c801514a96b4abaa8b1b2c330/sys/random/fortuna/fortuna.c#L139) piece of code.

The problem is that Fortuna expects some random events to initialize the pools. And that is something we don't do (yet). So I'm looking into other Fortuna implementations how they handle that, without deviating too much from the original implementation.

Last November was the last time I looked into this, so I have to get things sharp again :-)

basilfx commented:
@PeterKietzmann I have posted a question [here](https://crypto.stackexchange.com/questions/56390/seeding-the-fortuna-prng-for-the-first-time) for the remaining issue I have, regarding 'first boot'.

basilfx commented:
I have finished work on this PR and it is ready for review. Since a lot has changed since the initial feedback, I have squashed all commits that were not applicable anymore.

This PRNG will still be memory hungry, and the number of pools matter. I default to 32 (paper suggested), but you will have to lower to ~10 for a MCU with at least 32 KiB.

To test this PR, use `tests/rng` and compile it with something like this: `USEMODULE=prng_fortuna [CFLAGS="-DFORTUNA_POOLS=10"] make` (`CFLAGS` not needed for native).

smlng commented:
building this for SAMR21-XPRO fails: 
```
BOARD=samr21-xpro USEMODULE=prng_fortuna make all
Building application "tests_rng" for "samr21-xpro" with MCU "samd21".

[...]

/RIOT/tests/rng/bin/samr21-xpro/random.a(fortuna.o): In function `random_bytes':
/RIOT/sys/random/fortuna.c:102: multiple definition of `random_bytes'
/RIOT/tests/rng/bin/samr21-xpro/random.a(random.o):/RIOT/sys/random/random.c:44: first defined here
collect2: error: ld returned 1 exit status
make: *** [link] Error 1
```

also static-tests by travis complain about licence header stuff.

smlng commented on Makefile.dep:
```diff
@@ -5,7 +5,8 @@ OLD_USEPKG := $(sort $(USEPKG))
 
-# pull dependencies from drivers
+# pull dependencies from sys and drivers
+include $(RIOTBASE)/sys/Makefile.dep
```
not sure about introducing this here too. Though it is consistent with `/drivers`, but should still go in a separate PR and then adapt all sys modules deps, too.

basilfx replied:
I'm not sure what I should adapt. I could move it into a separate PR, but the only entry in `sys/Makefile.dep` is now Fortuna PRNG related. That would not make sense.

basilfx replied:
I've moved it to `sys/Makefile.dep` because of https://github.com/RIOT-OS/RIOT/pull/5675#discussion_r157827407.

smlng replied:
okay leave as is, I was just think that other sys-modules could be adapted, i.e., move deps form `RIOT/Makefile.dep` to `RIOT/sys/Makefile.dep` - but obviously not in this PR 😄

smlng commented on Makefile.dep:
```diff
@@ -607,6 +608,13 @@ ifneq (,$(filter random,$(USEMODULE)))
     USEMODULE += prng_tinymt32
   endif
 
+  ifneq (,$(filter prng_fortuna,$(USEMODULE)))
```
btw. with the above applied, this would go into `sys/Makefile.dep` right?

basilfx replied:
I think so. Not sure, because I basically duplicated `prng_tinymt32`.

smlng replied:
see comment above, leave as is for now and we adapt all sys-modules in a separate cleanup-PR.

smlng commented on sys/random/fortuna/fortuna.h:
```diff
@@ -0,0 +1,226 @@
+ * It is possible to use this method to clear an existing state.
+ *
+ * @param[in] state         PRNG state
+ * @return                  zero on succesful initialization
```
here and below, add empty line between param and return

basilfx replied:
Done.

smlng commented on sys/random/fortuna/fortuna.h:
```diff
@@ -0,0 +1,226 @@
+ * @return                  -2 on reading from unseeded PRNG
+ * @returned                -3 on other error
+ */
+int fortuna_random_data(fortuna_state_t *state, uint8_t *out, size_t bytes);
```
consider making `state` const if it is input param only.

basilfx replied:
Tried, but doesn't work since state is modified. I therefore changed it to `@param[inout]`.

smlng replied:
good

smlng commented on sys/random/fortuna/fortuna.h:
```diff
@@ -0,0 +1,226 @@
+ * @return                  -1 on zero bytes or more than 32 bytes
+ * @return                  -2 on invalid pool number
+ */
+int fortuna_add_random_event(fortuna_state_t *state, uint8_t *data,
```
same here (and below) consider making input params (pointers) const if possible.

basilfx replied:
Same.

basilfx commented:
@smlng Thanks for the review. I have added additional commits to fix the issues.

smlng commented on sys/random/fortuna.c:
```diff
@@ -0,0 +1,98 @@
+{
+    do {
+        /* read at most chunk bytes to not exhaust the state at once */
+        size_t chunk = bytes < FORTUNA_RESEED_LIMIT ?
```
put condition in parentheses, maybe?

smlng commented on sys/random/fortuna.c:
```diff
@@ -0,0 +1,98 @@
+        size_t chunk = bytes < FORTUNA_RESEED_LIMIT ?
+                       bytes : FORTUNA_RESEED_LIMIT;
+
+        int res = fortuna_random_data(&fortuna_state, out, bytes);
```
shouldn't this use `chunk` instead of `bytes`, i.e., `fortuna_random_data(&fortuna_state, out, chunk)`?

basilfx replied:
Yes, good catch!

basilfx commented:
@smlng Rebased and fixed the two comments.

smlng commented:
please squash, and the we give murdock a try

basilfx commented:
Murdock is happy.
